server/protocols/jmap-draft/src/main/java/org/apache/james/jmap/draft/methods/SetMailboxesUpdateProcessor.java

```diff
@@ -266,14 +269,20 @@ private void updateMailbox(Mailbox mailbox, MailboxUpdateRequest updateRequest,
                     .toMailboxAcl(),
                 mailboxSession);
         }
-        if (!originMailboxPath.equals(destinationMailboxPath)) {
-            mailboxManager.renameMailbox(mailbox.getId(), destinationMailboxPath, mailboxSession);
 
-            subscriptionManager.unsubscribe(mailboxSession, originMailboxPath.getName());
-            subscriptionManager.subscribe(mailboxSession, destinationMailboxPath.getName());
+        if (!originMailboxPath.equals(destinationMailboxPath)) {
+            List<MailboxRenamedResult> mailboxRenamedResults = mailboxManager.renameMailbox(mailbox.getId(), destinationMailboxPath, mailboxSession);
```

Comment:
the whole sequence is brittle, not only in term of implementation but also in term of design: by doing 3 API calls, you don't let the backend the opportunity to deal with failure properly.
What about having an API that allows to switch the subscription with the rename at the manager level instead?

Reply:
I think `subscription` should be handle right after rename `mailBox` completed.
Im not sure trigger call another `subscription api` or call directly here are better?
WDYT?

Reply:
I'm thinking about offering a rename API that handles subscription change (we could pass it an option to ask for that behavior).
I know that `SubscriptionManager` is a separated interface but having to take care of consequences of a rename "from the outside" looks like a design failure to me.
If we want to keep both interfaces (`MailboxManager` and `SubscriptionManager`) we can have a `MailboxListener` that takes care of the propagation of the subscription change.

Reply:
Matthieu is proposing to add a `SubscriptionManager.updateSubscription(oldName, newName, session)` method so that for example we can wrap this in a transaction 

(and I agree with this idea).

Reply:
So we let `SubscriptionManager` do the work - query get list mailBox related and handle it which mean we query twice for this data?

Reply:
We just move "unsubscribe + subscribe" login behind a method of SubscriptionManager interface.

Reply:
> I'm thinking about offering a rename API that handles subscription change (we could pass it an option to ask for that behavior).

No not really. 

More something like 

```
mailboxManager.renameMailbox(from, to, Option.RENAME_SUBSCRIPTIONS, session);
mailboxManager.renameMailbox(from, to, Option.KEEP_SUBSCRIPTIONS, session);

&& of course the id counterpart:

mailboxManager.renameMailbox(id, to, Option.RENAME_SUBSCRIPTIONS, session);
mailboxManager.renameMailbox(id, to, Option.KEEP_SUBSCRIPTIONS, session);
```

@mbaechler would you agree on delivering today work as a first version, and we let @nvduc91 work straight away on API refinement?

Reply:
> @mbaechler would you agree on delivering today work as a first version

What does merging it give us? My answer directly depends on this. Merging something just to move forward is IMO not a good habit. If we'll  make use of it while writing a new version, of course we can merge it.

Reply:
I think we should rework on that.